File: mobile_apps/Android/store/dropostore/src/main/java/com/dropo/store/models/datamodel/CartProductItems.java
package com.dropo.store.models.datamodel;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class CartProductItems implements Parcelable {


    public static final Creator<CartProductItems> CREATOR = new Creator<CartProductItems>() {
        @Override
        public CartProductItems createFromParcel(Parcel source) {
            return new CartProductItems(source);
        }

        @Override
        public CartProductItems[] newArray(int size) {
            return new CartProductItems[size];
        }
    };
    @SerializedName("tax")
    @Expose
    private double tax;
    @SerializedName("total_price")
    @Expose
    private double totalPrice;
    @SerializedName("total_tax")
    @Expose
    private double totalTax;
    @SerializedName("total_specification_tax")
    @Expose
    private double totalSpecificationTax;
    @SerializedName("item_tax")
    @Expose
    private double itemTax;
    @SerializedName("total_item_tax")
    @Expose
    private double totalItemTax;
    @SerializedName("note_for_item")
    private String itemNote;
    @SerializedName("total_item_price")
    private double totalItemAndSpecificationPrice;
    @SerializedName("item_details")
    private Item productItemsItem;
    @SerializedName("details")
    @Expose
    private String details;
    @SerializedName("item_name")
    @Expose
    private String itemName;
    @SerializedName("item_price")
    @Expose
    private double itemPrice;
    @SerializedName("total_specification_price")
    @Expose
    private double totalSpecificationPrice;
    @SerializedName("quantity")
    @Expose
    private int quantity;
    @SerializedName("item_id")
    @Expose
    private String itemId;
    @SerializedName("specifications")
    @Expose
    private List<ItemSpecification> specifications;
    @SerializedName("unique_id")
    @Expose
    private int uniqueId;
    @SerializedName("image_url")
    @Expose
    private List<String> imageUrl;

    @SerializedName("tax_details")
    @Expose
    private ArrayList<TaxesDetail> taxesDetails;

    public CartProductItems() {
    }

    protected CartProductItems(Parcel in) {
        this.totalPrice = in.readDouble();
        this.totalTax = in.readDouble();
        this.totalSpecificationTax = in.readDouble();
        this.itemTax = in.readDouble();
        this.totalItemTax = in.readDouble();
        this.itemNote = in.readString();
        this.totalItemAndSpecificationPrice = in.readDouble();
        this.productItemsItem = in.readParcelable(Item.class.getClassLoader());
        this.details = in.readString();
        this.itemName = in.readString();
        this.itemPrice = in.readDouble();
        this.totalSpecificationPrice = in.readDouble();
        this.quantity = in.readInt();
        this.itemId = in.readString();
        this.specifications = in.createTypedArrayList(ItemSpecification.CREATOR);
        this.uniqueId = in.readInt();
        this.imageUrl = in.createStringArrayList();
        taxesDetails = in.createTypedArrayList(TaxesDetail.CREATOR);
    }

    public double getTax() {
        return tax;
    }

    public void setTax(double tax) {
        this.tax = tax;
    }

    public double getTotalPrice() {
        return totalPrice;
    }

    public void setTotalPrice(double totalPrice) {
        this.totalPrice = totalPrice;
    }

    public double getTotalTax() {
        return totalTax;
    }

    public void setTotalTax(double totalTax) {
        this.totalTax = totalTax;
    }

    public double getTotalSpecificationTax() {
        return totalSpecificationTax;
    }

    public void setTotalSpecificationTax(double totalSpecificationTax) {
        this.totalSpecificationTax = totalSpecificationTax;
    }

    public double getItemTax() {
        return itemTax;
    }

    public void setItemTax(double itemTax) {
        this.itemTax = itemTax;
    }

    public double getTotalItemTax() {
        return totalItemTax;
    }

    public void setTotalItemTax(double totalItemTax) {
        this.totalItemTax = totalItemTax;
    }

    public String getItemNote() {
        return itemNote;
    }

    public void setItemNote(String itemNote) {
        this.itemNote = itemNote;
    }

    public int getQuantity() {
        return quantity;
    }

    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }

    public String getItemId() {
        return itemId;
    }

    public void setItemId(String itemId) {
        this.itemId = itemId;
    }

    public List<ItemSpecification> getSpecifications() {
        return specifications;
    }

    public void setSpecifications(List<ItemSpecification> specifications) {
        this.specifications = specifications;
    }

    public String getDetails() {
        return details;
    }

    public void setDetails(String details) {
        this.details = details;
    }

    public String getItemName() {
        return itemName;
    }

    public void setItemName(String itemName) {
        this.itemName = itemName;
    }

    public double getItemPrice() {
        return itemPrice;
    }

    public void setItemPrice(double itemPrice) {
        this.itemPrice = itemPrice;
    }

    public double getTotalSpecificationPrice() {
        return totalSpecificationPrice;
    }

    public void setTotalSpecificationPrice(double totalSpecificationPrice) {
        this.totalSpecificationPrice = totalSpecificationPrice;
    }

    public double getTotalItemAndSpecificationPrice() {
        return totalItemAndSpecificationPrice;
    }

    public void setTotalItemAndSpecificationPrice(double totalItemAndSpecificationPrice) {
        this.totalItemAndSpecificationPrice = totalItemAndSpecificationPrice;
    }

    public int getUniqueId() {
        return uniqueId;
    }

    public void setUniqueId(int uniqueId) {
        this.uniqueId = uniqueId;
    }

    public List<String> getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(List<String> imageUrl) {
        this.imageUrl = imageUrl;
    }

    public Item getProductItemsItem() {
        return productItemsItem;
    }

    public void setProductItemsItem(Item productItemsItem) {
        this.productItemsItem = productItemsItem;
    }

    public ArrayList<TaxesDetail> getTaxesDetails() {
        if (taxesDetails == null) {
            return new ArrayList<>();
        } else {
            return taxesDetails;
        }
    }

    public void setTaxesDetails(ArrayList<TaxesDetail> taxesDetails) {
        this.taxesDetails = taxesDetails;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeDouble(this.totalPrice);
        dest.writeDouble(this.totalTax);
        dest.writeDouble(this.totalSpecificationTax);
        dest.writeDouble(this.itemTax);
        dest.writeDouble(this.totalItemTax);
        dest.writeString(this.itemNote);
        dest.writeDouble(this.totalItemAndSpecificationPrice);
        dest.writeParcelable(this.productItemsItem, flags);
        dest.writeString(this.details);
        dest.writeString(this.itemName);
        dest.writeDouble(this.itemPrice);
        dest.writeDouble(this.totalSpecificationPrice);
        dest.writeInt(this.quantity);
        dest.writeString(this.itemId);
        dest.writeTypedList(this.specifications);
        dest.writeInt(this.uniqueId);
        dest.writeStringList(this.imageUrl);
        dest.writeTypedList(taxesDetails);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tax, totalPrice, totalTax, totalSpecificationTax, itemTax, totalItemTax, itemNote, totalItemAndSpecificationPrice, productItemsItem, details, itemName, itemPrice, totalSpecificationPrice, quantity, itemId, specifications, uniqueId, imageUrl, taxesDetails);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CartProductItems that = (CartProductItems) o;
        return uniqueId == that.uniqueId && itemId.equals(that.itemId) && specifications.equals(that.specifications);
    }
}

File: mobile_apps/IOS/user/Edelivery/Models/DataModel/Cart/CartProductItems.swift

import Foundation
 

public class CartProductItems {
	public var item_id : String?
	public var unique_id : Int?
	public var quantity : Int!
	public var item_name : String?
	public var image_url : Array<String>?
	public var item_price : Double?
    public var itemTax : Double = 0.0
    public var tax : Double = 0.0
    public var totalItemTax : Double = 0.0
	public var details : String?
	public var total_specification_price : Double?
    public var totalItemPrice:Double?
	public var specifications : [Specifications] = []
    public var producuItemsItem:ProductItemsItem?
    public var noteForItem:String! = ""
    public var maxItemQuantity:Int = 10
    public var totalSpecificationTax:Double = 0.0
    public var totalPrice:Double = 0.0
    public var totalTax:Double = 0.0
    public var taxDetails : [TaxesDetail]!

    public class func modelsFromDictionaryArray(array:NSArray) -> [CartProductItems] {
        var models:[CartProductItems] = []
        for item in array {
            models.append(CartProductItems(dictionary: item as! NSDictionary)!)
        }
        return models
    }
    
	required public init?(dictionary: NSDictionary) {

		item_id = (dictionary["item_id"] as? String) ?? ""
		unique_id = (dictionary["unique_id"] as? Int) ?? 0
		quantity = (dictionary["quantity"] as? Int) ?? 1
		item_name = (dictionary["item_name"] as? String) ?? ""
		item_price = (dictionary["item_price"] as? Double)?.roundTo()
        tax = ((dictionary["tax"] as? Double)?.roundTo()) ?? 0.0
        itemTax = ((dictionary["item_tax"] as? Double)?.roundTo()) ?? 0.0
        totalItemTax = ((dictionary["total_item_tax"] as? Double)?.roundTo()) ?? 0.0
        totalSpecificationTax =  ((dictionary["total_specification_tax"] as? Double)?.roundTo()) ?? 0.0
        producuItemsItem = dictionary["item_details"] as? ProductItemsItem
        totalItemPrice = (dictionary["total_item_price"] as? Double)?.roundTo() ?? 0.00
        maxItemQuantity = (dictionary["max_item_quantity"] as? Int) ?? 10
		details = dictionary["details"] as? String
        totalPrice = (dictionary["total_price"] as? Double)?.roundTo() ?? 0.00
        totalTax = (dictionary["total_tax"] as? Double)?.roundTo() ?? 0.00
        
		total_specification_price = (dictionary["total_specification_price"] as? Double)?.roundTo() ?? 0.00
		if (dictionary["specifications"] != nil) {
            specifications = Specifications.modelsFromDictionaryArray(array: dictionary["specifications"] as! NSArray)
        }
        if (dictionary["item_details"] != nil) {
            producuItemsItem = ProductItemsItem(dictionary: dictionary["item_details"] as! NSDictionary)
        }
        if preferenceHelper.getIsLoadItemImage() {
            if  (dictionary["image_url"] != nil) {
                image_url = (dictionary["image_url"] as! NSArray) as? Array<String>
            }
        }else {
            image_url = []
        }
        noteForItem = (dictionary["note_for_item"] as? String) ?? ""
        taxDetails = [TaxesDetail]()
        if let taxDetailsArray = dictionary["tax_details"] as? [[String:Any]]{
            for dic in taxDetailsArray{
                let value = TaxesDetail(fromDictionary: dic)
                taxDetails.append(value)
            }
        }
	}
    public init() {
    
    }

	public func dictionaryRepresentation() -> NSDictionary {

		let dictionary = NSMutableDictionary()

		dictionary.setValue(self.item_id, forKey: "item_id")
		dictionary.setValue(self.unique_id, forKey: "unique_id")
		dictionary.setValue(self.quantity, forKey: "quantity")
		dictionary.setValue(self.item_name, forKey: "item_name")
		dictionary.setValue(self.image_url, forKey: "image_url")
		dictionary.setValue(self.item_price, forKey: "item_price")
		dictionary.setValue(self.details, forKey: "details")
		dictionary.setValue(self.total_specification_price, forKey: "total_specification_price")
        dictionary.setValue(self.totalItemPrice, forKey: "total_item_price")
        dictionary.setValue(self.totalSpecificationTax, forKey: "total_specification_tax")
        dictionary.setValue(noteForItem, forKey: "note_for_item")
        dictionary.setValue(totalPrice, forKey: "total_price")
        dictionary.setValue(totalTax, forKey: "total_tax")
        dictionary.setValue(totalItemTax, forKey: "total_item_tax")
        dictionary.setValue(tax, forKey: "tax")
        
        if (self.producuItemsItem != nil) {
            dictionary.setValue(self.producuItemsItem, forKey: "item_details")
        }
        var myArray:[Any] = []
        for item in self.specifications {
            myArray.append(item.dictionaryRepresentation())
        }
        
        dictionary.setValue(myArray, forKey: "specifications")
        
        dictionary.setValue(self.maxItemQuantity, forKey: "max_item_quantity")
        if taxDetails != nil{
            var dictionaryElements = [[String:Any]]()
            for taxDetailsElement in taxDetails {
                dictionaryElements.append(taxDetailsElement.toDictionary())
            }
            dictionary["tax_details"] = dictionaryElements
        }
		return dictionary
	}
    
    public func getProductJson() -> NSDictionary {

        let dictionary = NSMutableDictionary()

        dictionary.setValue(self.item_id, forKey: "item_id")
        dictionary.setValue(self.unique_id, forKey: "unique_id")

        var myArray:[Any] = []
        for item in self.specifications {
            myArray.append(item.getProductJson())
        }
        
        dictionary.setValue(myArray, forKey: "specifications")
        
        return dictionary
    }

}
